src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/convert_moe_matmuls.cpp

```diff
@@ -0,0 +1,303 @@
+// Copyright (C) 2018-2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "convert_moe_matmuls.hpp"
+
+#include <initializer_list>
+#include <memory>
+#include <string>
+
+#include "itt.hpp"
+#include "openvino/cc/pass/itt.hpp"
+#include "openvino/core/graph_util.hpp"
+#include "openvino/core/node.hpp"
+#include "openvino/core/node_output.hpp"
+#include "openvino/core/rt_info.hpp"
+#include "openvino/core/type/element_type.hpp"
+#include "openvino/op/add.hpp"
+#include "openvino/op/broadcast.hpp"
+#include "openvino/op/clamp.hpp"
+#include "openvino/op/constant.hpp"
+#include "openvino/op/matmul.hpp"
+#include "openvino/op/minimum.hpp"
+#include "openvino/op/multiply.hpp"
+#include "openvino/op/reduce_sum.hpp"
+#include "openvino/op/reshape.hpp"
+#include "openvino/op/scatter_elements_update.hpp"
+#include "openvino/op/shape_of.hpp"
+#include "openvino/op/slice.hpp"
+#include "openvino/op/swish.hpp"
+#include "openvino/op/tile.hpp"
+#include "openvino/op/transpose.hpp"
+#include "openvino/op/unsqueeze.hpp"
+#include "openvino/pass/matcher_pass.hpp"
+#include "openvino/pass/pattern/matcher.hpp"
+#include "openvino/pass/pattern/op/label.hpp"
+#include "openvino/pass/pattern/op/optional.hpp"
+#include "openvino/pass/pattern/op/pattern.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "transformations/cpu_opset/common/op/batch_gather_matmul.hpp"
+
+using namespace ov::pass;
+namespace {
+// Note: intermediate nodes remain unchanged,
+// but we need to explicitly call shape inference for them to keep shapes consistency
+void validate_nodes(const pattern::PatternValueMap& map, const std::initializer_list<std::shared_ptr<ov::Node>> nodes) {
+    for (const auto& node : nodes) {
+        map.at(node).get_node_shared_ptr()->validate_and_infer_types();
+    }
+};
+
+std::shared_ptr<ov::op::v0::Unsqueeze> introduce_n_experts_dim(const ov::Output<ov::Node>& data) {
+    auto zero_const = std::make_shared<ov::op::v0::Constant>(ov::element::i32, ov::Shape{}, 0);
+    auto unsqueeze = std::make_shared<ov::op::v0::Unsqueeze>(data, zero_const);
+    ov::copy_runtime_info(data.get_node_shared_ptr(), {unsqueeze, zero_const});
+    return unsqueeze;
+}
+}  // namespace
+
+ov::intel_cpu::MoE2GeMM::MoE2GeMM() {
```

Comment:
I think such transformations are usually called "Fusion" in ov
I.e. BatchGatherMatmulsFusion